src/AvoStorage.ts

```diff
@@ -1,57 +1,112 @@
+  // onItemsFromLastSessionLoadedFuncs: Array<() => void> = [];
 
   constructor(shouldLog: boolean) {
     this.shouldLog = shouldLog;
     if (!process.env.BROWSER) {
       this.reactNative = require("react-native");
       this.Platform = this.reactNative.Platform;
       this.AsyncStorage = this.reactNative.AsyncStorage;
-      
 
       if (this.Platform.OS === "android") {
-        this.AsyncStorage.getAllKeys().then((keys: Array<any>) =>
-          this.AsyncStorage.multiGet(keys).then(
-            (keyVals: Array<Array<string>>) => {
-              keyVals.forEach((keyVal) => {
-                let key = keyVal[0];
-                this.data[key] = keyVal[1];
-              });
-              this.initialized = true;
-              this.onInitFuncs.forEach((func) => {
-                func();
-              });
-            }
-          )
-        );
+        this.loadAndroidDataToMemoryToAvoidAsyncQueries(() => {
+          this.initializeStorageAndroid();
+        });
       } else {
-        this.initialized = true;
+        this.initializeStorageIos();
       }
     } else {
-      this.initialized = true;
+      this.initializeStorageWeb(this.isLocalStorageAvailable());
     }
   }
 
-  runOnInit(func: () => void) {
-    if (this.initialized === true) {
+  loadAndroidDataToMemoryToAvoidAsyncQueries(onLoaded: () => void) {
+    this.AsyncStorage.getAllKeys().then((keys: Array<any>) =>
+      this.AsyncStorage.multiGet(keys).then((keyVals: Array<Array<string>>) => {
+        keyVals.forEach((keyVal) => {
+          let key = keyVal[0];
+          this.androidMemoryDataToAvoidAsyncQueries[key] = keyVal[1];
+        });
+        onLoaded();
+      })
+    );
+  }
+
+  isLocalStorageAvailable(): boolean {
+    const uid = new Date().toISOString();
+    try {
+      window.localStorage.setItem(uid, uid);
+      if (window.localStorage.getItem(uid) === uid) {
+        window.localStorage.removeItem(uid);
+        return true;
+      } else {
+        return false;
+      }
+    } catch (error) {
+      return false;
+    }
+  }
+
+  initializeStorageAndroid() {
+    this.storageInitialized = true;
+    this.onStorageInitFuncs.forEach((func) => {
+      func();
+    });
+  }
+
+  initializeStorageIos() {
+    this.storageInitialized = true;
+    this.onStorageInitFuncs.forEach((func) => {
+      func();
+    });
+  }
+
+  initializeStorageWeb(isLocalStorageAvailable: boolean) {
+    this.storageInitialized = true;
+    if (isLocalStorageAvailable === false) {
+      this.useFallback = true;
+    }
+    this.onStorageInitFuncs.forEach((func) => {
+      func();
+    });
+  }
+
+  runOnStorageInit(func: () => void) {
+    if (this.storageInitialized === true) {
       func();
     } else {
-      this.onInitFuncs.push(func);
+      this.onStorageInitFuncs.push(func);
     }
   }
 
   getItemAsync<T>(key: string): Promise<T | null> {
     let maybeItem;
-    if (process.env.BROWSER) {
+    if (this.useFallback === true) {
```

Comment:
The only logical difference with your previous implementation is that here we don't do initialized check. I think it should be just fine, what do you think @logason ?

Reply:
I think I'd like to keep the initialize check in place, to guard us from the case where a storage method is being called before initialization has been triggered (should never happen, but I want to 100% guarantee us from throwing errors here)

Reply:
I see... I think the proper implementation here would be to return a promise that is resolved after the storage is ready then. And on android it's right away, even before storageInitialized flag is set, because of the async nature of the async storage, but in browser it should wait for storageInitialized flag.

Reply:
That's pretty tricky to implement though, I say we setup the guard for now to prevent a crash and keep the async for later. We could wrap the logic that calls the async function in the onInit callback to make sure it only fires when the storage is ready?

Reply:
Just adding a guard will break android, I think I got the promise approach, just pushed and tested in the example app, getting valid data with the getItemAsync method.